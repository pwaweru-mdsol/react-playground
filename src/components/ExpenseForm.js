import React, { Component } from "react";
import { SingleDatePicker } from "react-dates";
import "react-dates/initialize";
import moment from "moment";

const ERROR_MESSAGE = "Please enter a description or an amount";

export class ExpenseForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '' ,
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : '',
      calenderFocused: false,
      error: undefined
    }
  }

  onDescriptionChange = event => {
    const description = event.target.value;
    this.setState(() => ({ description }));
  };

  onNoteChange = event => {
    const note = event.target.value;
    this.setState(() => ({ note }));
  };

  onAmountChange = event => {
    const amount = event.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onDateChange = createdAt => {
    if (createdAt) {
      this.setState(() => ({ createdAt: createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calenderFocused: focused }));
  };

  onSubmit = event => {
    event.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({ error: ERROR_MESSAGE }));
    } else {
      this.setState(() => ({ error: undefined }));
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      });
    }
  };

  render() {
    return (
      <div>
        {this.state.error ? <p>{ERROR_MESSAGE}!</p> : null}
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="number"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />

          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
          />
          <textarea
            placeholder="Add a note for your exepnse (optional)"
            onChange={this.onNoteChange}
          />
          <button>Submit Expense</button>
        </form>
      </div>
    );
  }
}

export default ExpenseForm;
