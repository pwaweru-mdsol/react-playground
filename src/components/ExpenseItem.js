import React from "react";
import { removeExpense } from "../actions/expenses";
import { connect } from "react-redux";

const ExpenseItem = ({ dispatch, uuid, description, amount, createdAt }) => (
  <div>
    <h3>{description}</h3>
    <p>
      Â£{amount} - {createdAt}
    </p>
    <button
      onClick={() => {
        dispatch(removeExpense(uuid));
      }}
    >
      remove
    </button>
  </div>
);

export default connect()(ExpenseItem);
