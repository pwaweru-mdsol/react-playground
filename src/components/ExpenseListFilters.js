import React, { Component } from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByDate,
  sortByAmount,
  setStartDate,
  setEndDate
} from "../actions/filters";

import uuid from "uuid/v4";

import { DateRangePicker } from "react-dates";

export class ExpenseListFilters extends Component {
  state = {
    calenderFocused: null
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = calenderFocused => {
    this.setState(() => ({ calenderFocused }));
  };

  onTextChange = event => {
    this.props.setTextFilter(event.target.value);
  }

  onSortChange = event => {
    if(event.target.value == 'date') {
      this.props.sortByDate()
    } else if(event.target.value == 'amount') {
      this.props.sortByAmount()
    }
  }

  render() {
    return (
      <div>
        <input
          type="text"
          value={this.props.filters.text}
          onChange={this.onTextChange}
        />

        <select onChange={this.onSortChange}>
          <option value="date">Date</option>
          <option value="amount">Amount</option>
        </select>

        <DateRangePicker
          startDateId={uuid()}
          endDateId={uuid()}
          startDate={this.props.filters.startDate}
          endDate={this.props.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calenderFocused}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          showClearDates={true}
          isOutsideRange={() => false}
        />
      </div>
    );
  }
}

const mapStateToProps = state => {
  return {
    filters: state.filters
  };
};

const mapDispatchToPropes = dispatch => ({
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount()),
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: setEndDate => dispatch(setEndDate(setEndDate)),
});

export default connect(mapStateToProps, mapDispatchToPropes)(ExpenseListFilters);
