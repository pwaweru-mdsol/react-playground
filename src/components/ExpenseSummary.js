import React from "react";
import { connect } from "react-redux";
import currency from 'format-currency';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense' : 'expenses';
  const formattedCurrency = currency(expenseTotal / 100);
  return ( 
    <p>
      Viewing {expenseCount} {expenseWord} totalling ${formattedCurrency}
    </p>
  );
};

const mapStateToProps = state => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  console.log('visibleExpenses: ', visibleExpenses);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: selectExpensesTotal(visibleExpenses)
  };
};


export default connect(mapStateToProps)(ExpenseSummary);