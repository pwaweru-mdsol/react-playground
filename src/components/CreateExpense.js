import React, { Component } from "react";
import { connect } from 'react-redux';
import ExpenseForm from "./ExpenseForm";
import { addExpense } from '../actions/expenses';

export class CreateExpense extends Component {
  onSubmit = expense => {
    this.props.addExpense(expense);
    this.props.history.push('/');
  }

  render(){
    return(
      <div>
      Create Expense!
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
    </div>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  addExpense: expense => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(CreateExpense);
