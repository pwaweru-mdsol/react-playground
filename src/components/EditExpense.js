import React, { Component } from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpense extends Component {
  onSubmit = expense => {
    this.props.editExpense(this.props.expense.uuid, expense);
    this.props.history.push('/');
  }
   onRemove = expense => {
    this.props.removeExpense(this.props.expense.uuid);
    this.props.history.push('/');
   }

   render() {
     return (
        <div>
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
        />
        <button
          onClick={this.onRemove}
        >
           remove
        </button>
      </div>
     )
   }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(expense => expense.uuid === props.match.params.uuid)
});

const mapDispatchToProps = (dispatch, props) => ({
  editExpense: (uuid, expense) => dispatch(editExpense(uuid, expense)),
  removeExpense: uuid => dispatch(removeExpense(uuid)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpense);
