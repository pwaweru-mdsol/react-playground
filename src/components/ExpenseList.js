import React from 'react';
import { connect } from 'react-redux';

import ExpenseItem from './ExpenseItem';

const ExpenseList = (props) => (
	<div>
		<h1>Expense List</h1>
		{
			props.expenses.map(expense => (
				<ExpenseItem description={expense.description} amount={expense.amount} createdAt={expense.createdAt}/>
			))
		}
	</div>
);

const mapStateToProps = (state => {
	return { expenses: state.expenses, filters: state.filters }
});

export default connect(mapStateToProps)(ExpenseList);		