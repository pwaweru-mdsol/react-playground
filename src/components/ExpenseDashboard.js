import React from "react";
import ExpenseList from "./ExpenseList";
import ExpenseListFilters from "./ExpenseListFilters";
import ExpenseSummary from './ExpenseSummary';

const ExpenseDashboard = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseSummary />
    <ExpenseList />
  </div>
);

export default ExpenseDashboard;
