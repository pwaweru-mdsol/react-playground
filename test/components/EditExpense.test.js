import React from 'react';
import { shallow } from 'enzyme';
import { EditExpense } from '../../src/components/EditExpense';

const expenses = [
    {
      uuid: "1",
      description: "hello world",
      note: "",
      amount: 1.95,
      createdAt: '1'
    }
];

describe('EditExpense', () => {
  let wrapper;
  let editExpenseSpy = jest.fn();
  let removeExpenseSpy = jest.fn();
  let historySpy = { push: jest.fn() }

  beforeEach(() => {
    wrapper = shallow(<EditExpense 
      history={historySpy}
      editExpense={editExpenseSpy} 
      removeExpense={removeExpenseSpy} 
      expense={expenses[0]}
    /> );
      
  });

  test('should render EditExpense component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should handle onSubmit', () => {
    wrapper.find('ExpenseForm').prop('onSubmit')(expenses[0]);
    expect(editExpenseSpy).toHaveBeenCalledWith(expenses[0].uuid, expenses[0]);
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
  });

  test('should handle onRemove', () => {
    wrapper.find('button').simulate('click');
    expect(removeExpenseSpy).toHaveBeenCalledWith(expenses[0].uuid)
    expect(historySpy.push).toHaveBeenLastCalledWith('/');
  });
})

