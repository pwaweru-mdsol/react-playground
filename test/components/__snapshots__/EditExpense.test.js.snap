// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpense should render EditExpense component 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 1.95,
        "createdAt": "1",
        "description": "hello world",
        "note": "",
        "uuid": "1",
      }
    }
    onSubmit={[Function]}
  />
  <button
    onClick={[Function]}
  >
    remove
  </button>
</div>
`;
